python/ray/tune/ray_trial_executor.py

```diff
@@ -87,73 +88,38 @@ def unwrap(self):
-        if force_cleanup < 0:
-            force_cleanup = 0
-        self._force_cleanup = force_cleanup
-
-    def add(self, trial: Trial, actor: ActorHandle):
-        """Adds a trial actor to be stopped.
-
-        If the number of futures exceeds the threshold, the cleanup mechanism
-        will kick in.
-
-        Args:
-            trial (Trial): The trial corresponding to the future.
-            actor (ActorHandle): Handle to the trainable to be stopped.
-        """
-        future = actor.stop.remote()
+    def __init__(self, queue: Queue):
+        super().__init__()
+        self._futures = {}
+        self._queue = queue
+        self._should_finish = False
 
+    def _process(self, item):
+        if not item:
+            self._should_finish = True
+            self._queue.task_done()
+            return
+        actor, pg = item
+        self._futures[actor.stop.remote()] = pg
         del actor
 
-        self._cleanup_map[future] = trial
-        if len(self._cleanup_map) > self.threshold:
-            self.cleanup(partial=True)
+    def run(self):
+        while True:
+            if self._should_finish:
+                return
+            item = self._queue.get(block=len(self._futures) == 0)
+            self._process(item)
+            while not self._queue.empty():
+                item = self._queue.get()
+                self._process(item)
 
-    def cleanup(self, partial: bool = True):
-        """Waits for cleanup to finish.
-
-        If partial=False, all futures are expected to return. If a future
-        does not return within the timeout period, the cleanup terminates.
-        """
-        # At this point, self._cleanup_map holds the last references
-        # to actors. Removing those references either one-by-one
-        # (graceful termination case) or all at once, by reinstantiating
-        # self._cleanup_map (forceful termination case) will cause Ray
-        # to kill the actors during garbage collection.
-        logger.debug("Cleaning up futures")
-        num_to_keep = int(self.threshold) / 2 if partial else 0
-        while len(self._cleanup_map) > num_to_keep:
-            dones, _ = ray.wait(
-                list(self._cleanup_map),
-                timeout=DEFAULT_GET_TIMEOUT
-                if not self._force_cleanup else self._force_cleanup)
-            if not dones:
-                logger.warning(
-                    "Skipping cleanup - trainable.stop did not return in "
-                    "time. Consider making `stop` a faster operation.")
-                if not partial and self._force_cleanup:
-                    logger.warning(
-                        "Forcing trainable cleanup by terminating actors.")
-                    self._cleanup_map = {}
-                    return
-            else:
-                done = dones[0]
-                del self._cleanup_map[done]
+            ready_futures, _ = ray.wait(list(self._futures.keys()), timeout=0.5)
```

Comment:
I am wondering if there can be a combined wait for queue.get() and ray.get() at the same time. So that this random chosen 0.5 is not necessary anymore.